File: python/leetcode/walking-robot-simulation.py
class Solution:
    def robotSim(self, commands, obstacles):
        """
        :type commands: List[int]
        :type obstacles: List[List[int]]
        :rtype: int
        """
        c = 0
        directions, obstacles = [(0, 1), (1, 0), (0, -1), (-1, 0)], set(map(tuple, obstacles))
        dir_pos = 0
        x = 0
        y = 0
        max_distance = 0
        while c < len(commands):
            command = commands[c]
            if command == -1:
                dir_pos = (dir_pos + 1) % 4
            elif command == -2:
                dir_pos = (dir_pos - 1) % 4
            else:
                while command > 0:
                    nextX = x + directions[dir_pos][0]
                    nextY = y + directions[dir_pos][1]
                    nextCoord = (nextX, nextY)
                    if nextCoord in obstacles:
                        break
                    x = nextX
                    y = nextY
                    command -= 1
                max_distance = max(max_distance, x**2 + y**2)
            c += 1
        return max_distance

File: javascript/leetcode/walking-robot-simulation.js
/**
 * @param {number[]} commands
 * @param {number[][]} obstacles
 * @return {number}
 */
var robotSim = function(commands, obstacles) {
  let c = 0;
  let directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
  let dirPos = 0;
  let x = 0;
  let y = 0;
  let command, nextX, nextY;
  let maxDistance = 0;
  while (c < commands.length) {
    command = commands[c];
    if (command === -1 ) {
        dirPos = (dirPos + 1) % 4;
    } else if (command === -2) {
        dirPos = dirPos === 0 ? 3 : dirPos - 1;
    } else {
        while (command > 0) {
            nextX = x + directions[dirPos][0];
            nextY = y + directions[dirPos][1];
            nextCoord = [nextX, nextY];
            if (isObstacle(nextCoord, obstacles)) {
                console.log('boop')
                break;
            }
            x = nextX;
            y = nextY;
            command--;
        }
        maxDistance = Math.max(maxDistance, x**2 + y**2);
    }
    c++;
  }
  return maxDistance;
};

function isObstacle(nextCoord, obstacles) {
    for (let i = 0; i < obstacles.length; i++) {
        if (obstacles[i][0] === nextCoord[0] && obstacles[i][1] === nextCoord[1]) { return true; }
    }
    return false;
}
